python/drned_xmnr/action.py

```diff
@@ -83,6 +84,39 @@ def action_response(self, uinfo, result, output):
             output.failure = result['failure']
 
 
+class CompletionHandler(dp.Action):
+
+    # @dp.Action.completion
+    # wrapper does not exist in PyAPI at the time of this implementation
+    def cb_completion(self, uinfo, cli_style, token, completion_char,
+                      kp, cmdpath, cmdparam_id, simpleType, extra):
+        self.log.debug("========== drned_xmnr cb_completion() ==========")
+        self.log.debug("thandle={0} usid={1}".format(uinfo.actx_thandle,
+                                                     uinfo.usid))
+        try:
+            # TODO - escape/sanitize input "token"
+            pattern_to_match = str(token)
+            if '*' not in pattern_to_match:
+                pattern_to_match += '*'
+            self.log.debug('Find completions for: \"' + pattern_to_match + '\"')
+            matched_paths = glob.glob(pattern_to_match)
+            tv = []
+            for path in matched_paths:
+                tv.append((dp.COMPLETION, str(path), None))
+            self.log.debug('Found matches:' + str(tv))
```

Comment:
I'd be more careful with logging - the completion might be invoked several times for one CLI line, no need to debug-log really everything

Reply:
fixed in succeeding commit...